### DISCUSSION TITLE
Why services are reset at the start of the next request but not at the end of current request

### DISCUSSION THREAD

**Discussion by AdamKatzDev** [symfony/symfony#52196] [Q&A]
In case when the kernel is reused the services are reset at the start of the next request which might lead to unwanted effects.
For example I've implemented custom SQS transport that collects messages in batches of 10 to send them using `SendMessageBatch`. I rely on `ResetInterface` and `__destruct` to send the last batch. In cases when kernel is not reused (`mod_php`, `PHP-FPM`) `__destruct` handles sending the last batch. And Symfony Messenger Workers reset services after processing every message so everything works.
But when reusing kernel (e.g. when using Roadrunner) with current reset logic the last batch will be sent only at the start of next request, so:
- in worst case scenario (very low RPS per process) the last message batch can be delayed indefinitely
- next request will be affected by the latency caused by making `SendMessageBatch` request

So was it made that way on purpose or such scenario just wasn't considered at the moment this feature was implemented?

**UPD:** I understand why it was made that way. As @stof has written below it was made to avoid the overhead of resetting kernels that are dying anyway. But why not to move the reset to the end of requests and do nothing if the kernel is a dying one? Is this not possible to implement or prone to errors?

**Comment by stof** [comment]
This was done this way to avoid the overhead of resetting all services in case the kernel is not reused (which is the most common case for deployments handling requests as PHP-FPM is far more common than Roadrunner).

You might want to trigger the submission of your message batch in the `kernel.terminate` event.

**Reply by AdamKatzDev** [reply]
Right, I've figured that this is what I might have to do (good thing that this is not a real situation and just an observation).
What worries me the most is that the way resettable services work is a bit inconsistent between Requests and Symfony Messenger. And honestly this is not what I expected also.
